packages/table/src/table.tsx

```diff
@@ -1265,6 +1302,171 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         this.scrollBodyToFocusedCell(newFocusedCell);
     }
 
+    // no good way to call arrow-key keyboard events from tests
+    /* istanbul ignore next */
+    private handleFocusMoveInternal = (e: KeyboardEvent, direction: "up" | "down" | "left" | "right") => {
+        e.preventDefault();
+        e.stopPropagation();
+
+        const { focusedCell, selectedRegions } = this.state;
+        const { grid } = this;
+
+        if (focusedCell == null) {
+            // halt early if we have a selectedRegionTransform or something else in play that nixes
+            // the focused cell.
+            return;
+        }
+
+        let newFocusedCell = {
+            col: focusedCell.col,
+            focusSelectionIndex: focusedCell.focusSelectionIndex,
+            row: focusedCell.row,
+        };
+
+        // if we're not in any particular focus cell region, and one exists, go to the first cell of the first one
+        if (focusedCell.focusSelectionIndex == null && selectedRegions.length > 0) {
+            const focusCellRegion = Regions.getCellRegionFromRegion(
+                selectedRegions[0],
+                grid.numRows,
+                grid.numCols,
+            );
+
+            newFocusedCell = {
+                col: focusCellRegion.cols[0],
+                focusSelectionIndex: 0,
+                row:  focusCellRegion.rows[0],
+            };
+        } else {
+            if (selectedRegions.length === 0) {
+                this.handleFocusMove(e, direction);
+                return;
+            }
+
+            const focusCellRegion = Regions.getCellRegionFromRegion(
+                selectedRegions[focusedCell.focusSelectionIndex],
+                grid.numRows,
+                grid.numCols,
+            );
+
+            if (focusCellRegion.cols[0] === focusCellRegion.cols[1] &&
+                focusCellRegion.rows[0] === focusCellRegion.rows[1] &&
+                selectedRegions.length === 1) {
+                this.handleFocusMove(e, direction);
+                return;
+            }
+
+            switch (direction) {
+                case "up":
+                    newFocusedCell.row -= 1;
```

Comment:
I mean, I _could_ . 

This code is very... abstract, as you said, though. And abstracting this much makes it super hard to follow. 

I'll see if I can't find some middle ground, or something. Or possibly just comment it to death.